File: android/src/main/kotlin/com/utexo/rgb_sdk_flutter/RlnNodeStore.kt
package com.utexo.rgb_sdk_flutter

import org.utexo.rgblightningnode.NativeExternalSigner
import org.utexo.rgblightningnode.SdkNode

internal class RlnNodeStore {
    private val nodes = mutableMapOf<Long, SdkNode>()
    private val states = mutableMapOf<Long, NodeLifecycleState>()
    private val preUnlockStates = mutableMapOf<Long, NodeLifecycleState>()
    private val storageDirByNodeId = mutableMapOf<Long, String>()
    private var nextId = 1L
    private val signers = mutableMapOf<Long, NativeExternalSigner>()
    private var nextSignerId = 1L

    enum class NodeLifecycleState {
        CREATED,
        INITIALIZED,
        UNLOCKING,
        UNLOCKED,
        SHUTDOWN
    }

    @Synchronized
    fun ensureStorageAvailable(storageDirPath: String) {
        val existing = storageDirByNodeId.entries.firstOrNull {
            it.value == storageDirPath.trim()
        }?.key
        if (existing != null && states[existing] != NodeLifecycleState.SHUTDOWN) {
            throw RlnStateConflict("RLN storage is already owned by an active node")
        }
    }

    @Synchronized
    fun create(node: SdkNode, storageDirPath: String): Long {
        val normalizedPath = storageDirPath.trim()
        if (normalizedPath.isNotEmpty()) {
            val existingId = storageDirByNodeId.entries.firstOrNull { it.value == normalizedPath }?.key
            if (existingId != null) {
                if (states[existingId] == NodeLifecycleState.SHUTDOWN) {
                    remove(existingId)
                } else {
                    throw RlnStateConflict("RLN storage is already owned by an active node")
                }
            }
        }

        val id = nextId++
        nodes[id] = node
        states[id] = NodeLifecycleState.CREATED
        storageDirByNodeId[id] = normalizedPath
        return id
    }

    @Synchronized
    fun get(id: Long): SdkNode {
        return nodes[id] ?: throw NodeNotFound()
    }

    @Synchronized
    fun getState(id: Long): NodeLifecycleState {
        return states[id] ?: throw NodeNotFound()
    }

    @Synchronized
    fun markInitialized(id: Long) {
        when (val state = getState(id)) {
            NodeLifecycleState.CREATED -> states[id] = NodeLifecycleState.INITIALIZED
            NodeLifecycleState.INITIALIZED -> Unit
            NodeLifecycleState.UNLOCKED,
            NodeLifecycleState.SHUTDOWN -> throw RlnStateConflict(
                "RLN init is not allowed while node is in state: $state"
            )
            NodeLifecycleState.UNLOCKING -> throw RlnStateConflict(
                "Cannot initialize RLN node while unlock is in progress"
            )
        }
    }

    @Synchronized
    fun beginUnlock(id: Long): NodeLifecycleState {
        return when (val state = getState(id)) {
            NodeLifecycleState.CREATED,
            NodeLifecycleState.INITIALIZED,
            NodeLifecycleState.SHUTDOWN -> {
                preUnlockStates[id] = state
                states[id] = NodeLifecycleState.UNLOCKING
                NodeLifecycleState.UNLOCKING
            }
            NodeLifecycleState.UNLOCKED -> NodeLifecycleState.UNLOCKED
            NodeLifecycleState.UNLOCKING -> throw RlnStateConflict("RLN unlock is already in progress")
        }
    }

    @Synchronized
    fun markUnlocked(id: Long) {
        if (states.containsKey(id)) {
            states[id] = NodeLifecycleState.UNLOCKED
            preUnlockStates.remove(id)
        }
    }

    @Synchronized
    fun markShutdown(id: Long) {
        if (states.containsKey(id)) {
            states[id] = NodeLifecycleState.SHUTDOWN
        }
    }

    @Synchronized
    fun rollbackUnlock(id: Long) {
        if (states[id] == NodeLifecycleState.UNLOCKING) {
            states[id] = preUnlockStates.remove(id) ?: NodeLifecycleState.INITIALIZED
        }
    }

    @Synchronized
    fun remove(id: Long) {
        val node = nodes.remove(id)
        val state = states.remove(id)
        preUnlockStates.remove(id)
        storageDirByNodeId.remove(id)
        if (node != null) closeNode(node, state)
    }

    private fun closeNode(node: SdkNode, state: NodeLifecycleState?) {
        try {
            if (state != NodeLifecycleState.SHUTDOWN) node.shutdown()
        } finally {
            // Releasing the UniFFI handle alone is not a node shutdown contract.
            node.close()
        }
    }

    @Synchronized
    fun createSigner(signer: NativeExternalSigner): Long {
        val id = nextSignerId++
        signers[id] = signer
        return id
    }

    @Synchronized
    fun getSigner(id: Long): NativeExternalSigner {
        return signers[id] ?: throw SignerNotFound()
    }

    @Synchronized
    fun removeSigner(id: Long) {
        val signer = signers.remove(id)
        signer?.close()
    }

    @Synchronized
    fun clearAll() {
        val nodesToClose = nodes.map { (id, node) -> node to states[id] }
        val signersToClose = signers.values.toList()
        nodes.clear()
        states.clear()
        preUnlockStates.clear()
        storageDirByNodeId.clear()
        signers.clear()
        var firstFailure: Exception? = null
        nodesToClose.forEach { (node, state) ->
            try {
                closeNode(node, state)
            } catch (error: Exception) {
                firstFailure = firstFailure ?: error
            }
        }
        signersToClose.forEach { signer ->
            try {
                signer.close()
            } catch (error: Exception) {
                firstFailure = firstFailure ?: error
            }
        }
        firstFailure?.let { throw it }
    }

    @Synchronized
    fun snapshot(): RlnNodeStoreSnapshot {
        return RlnNodeStoreSnapshot(
            nodeCount = nodes.size,
            signerCount = signers.size,
            states = states.toMap(),
            storageDirByNodeId = storageDirByNodeId.toMap()
        )
    }
}

internal class NodeNotFound : IllegalStateException("Native node handle is unavailable.")
internal class SignerNotFound : IllegalStateException("Native signer handle is unavailable.")
internal class RlnStateConflict(message: String) : IllegalStateException(message)

internal data class RlnNodeStoreSnapshot(
    val nodeCount: Int,
    val signerCount: Int,
    val states: Map<Long, RlnNodeStore.NodeLifecycleState>,
    val storageDirByNodeId: Map<Long, String>
)

File: ios/Classes/RlnNodeStore.swift
import Foundation

final class RlnNodeStore {
  static let shared = RlnNodeStore()

  enum NodeLifecycleState: Equatable {
    case created
    case initialized
    case unlocking
    case unlocked
    case shutdown
  }

  private var nodes: [Int64: SdkNode] = [:]
  private var states: [Int64: NodeLifecycleState] = [:]
  private var preUnlockStates: [Int64: NodeLifecycleState] = [:]
  private var storageDirByNodeId: [Int64: String] = [:]
  private var nextId: Int64 = 1
  private var signers: [Int64: NativeExternalSigner] = [:]
  private var nextSignerId: Int64 = 1
  private let queue = DispatchQueue(label: "com.utexo.rgb_sdk_flutter.rln-node-store")

  private init() {}

  func create(node: SdkNode, storageDirPath: String) throws -> Int64 {
    try queue.sync {
      let normalizedPath = storageDirPath.trimmingCharacters(in: .whitespacesAndNewlines)
      if !normalizedPath.isEmpty,
         let existingId = storageDirByNodeId.first(where: { $0.value == normalizedPath })?.key {
        if states[existingId] == .shutdown {
          nodes.removeValue(forKey: existingId)?.shutdown()
          nodes[existingId] = node
          states[existingId] = .created
          return existingId
        }

        throw RlnStoreError.nodeAlreadyExists(storageDirPath: normalizedPath)
      }

      let id = nextId
      nextId += 1
      nodes[id] = node
      states[id] = .created
      storageDirByNodeId[id] = normalizedPath
      return id
    }
  }

  func get(id: Int64) throws -> SdkNode {
    try queue.sync {
      guard let node = nodes[id] else {
        throw RlnStoreError.nodeNotFound(id: id)
      }
      return node
    }
  }

  func getState(id: Int64) throws -> NodeLifecycleState {
    try queue.sync {
      guard let state = states[id] else {
        throw RlnStoreError.nodeNotFound(id: id)
      }
      return state
    }
  }

  func markInitialized(id: Int64) throws {
    try queue.sync {
      guard let state = states[id] else {
        throw RlnStoreError.nodeNotFound(id: id)
      }

      switch state {
      case .created:
        states[id] = .initialized
      case .initialized:
        break
      case .unlocked, .shutdown:
        throw RlnStoreError.invalidState("RLN init is not allowed while node is in state: \(state)")
      case .unlocking:
        throw RlnStoreError.invalidState("Cannot initialize RLN node while unlock is in progress")
      }
    }
  }

  func beginUnlock(id: Int64) throws -> NodeLifecycleState {
    try queue.sync {
      guard let state = states[id] else {
        throw RlnStoreError.nodeNotFound(id: id)
      }

      switch state {
      case .created, .initialized, .shutdown:
        preUnlockStates[id] = state
        states[id] = .unlocking
        return .unlocking
      case .unlocked:
        return .unlocked
      case .unlocking:
        throw RlnStoreError.invalidState("RLN unlock is already in progress")
      }
    }
  }

  func markUnlocked(id: Int64) {
    queue.sync {
      if states[id] != nil {
        states[id] = .unlocked
        preUnlockStates.removeValue(forKey: id)
      }
    }
  }

  func markShutdown(id: Int64) {
    queue.sync {
      if states[id] != nil {
        states[id] = .shutdown
      }
    }
  }

  func rollbackUnlock(id: Int64) {
    queue.sync {
      if states[id] == .unlocking {
        states[id] = preUnlockStates.removeValue(forKey: id) ?? .initialized
      }
    }
  }

  func remove(id: Int64) {
    queue.sync {
      let state = states[id]
      let node = nodes.removeValue(forKey: id)
      states.removeValue(forKey: id)
      preUnlockStates.removeValue(forKey: id)
      storageDirByNodeId.removeValue(forKey: id)
      if let node, state != .shutdown {
        node.shutdown()
      }
    }
  }

  func createSigner(_ signer: NativeExternalSigner) -> Int64 {
    queue.sync {
      let id = nextSignerId
      nextSignerId += 1
      signers[id] = signer
      return id
    }
  }

  func getSigner(id: Int64) throws -> NativeExternalSigner {
    try queue.sync {
      guard let signer = signers[id] else {
        throw RlnStoreError.signerNotFound(id: id)
      }
      return signer
    }
  }

  func removeSigner(id: Int64) {
    queue.sync {
      signers.removeValue(forKey: id)
    }
  }

  func clearAll() {
    queue.sync {
      let nodesToShutdown = Array(nodes.values)
      nodes.removeAll()
      states.removeAll()
      preUnlockStates.removeAll()
      storageDirByNodeId.removeAll()
      signers.removeAll()
      nodesToShutdown.forEach { $0.shutdown() }
    }
  }

  func snapshot() -> RlnNodeStoreSnapshot {
    queue.sync {
      RlnNodeStoreSnapshot(
        nodeCount: nodes.count,
        signerCount: signers.count,
        states: states,
        storageDirByNodeId: storageDirByNodeId
      )
    }
  }
}

struct RlnNodeStoreSnapshot {
  let nodeCount: Int
  let signerCount: Int
  let states: [Int64: RlnNodeStore.NodeLifecycleState]
  let storageDirByNodeId: [Int64: String]
}

enum RlnStoreError: Error, LocalizedError {
  case nodeAlreadyExists(storageDirPath: String)
  case nodeNotFound(id: Int64)
  case signerNotFound(id: Int64)
  case invalidState(String)

  var errorDescription: String? {
    switch self {
    case .nodeAlreadyExists(let storageDirPath):
      return "RLN node already exists for storageDirPath: \(storageDirPath)"
    case .nodeNotFound(let id):
      return "RLN node with id \(id) not found"
    case .signerNotFound(let id):
      return "Native signer with id \(id) not found"
    case .invalidState(let message):
      return message
    }
  }
}
